File: problems-adapt/0053_largest-subarray-sum/solution_kadane.cpp
class Solution {
  public:
    int largestSubarraySum(vector<int> &nums) {
        // Kadane's algorithm: current is the best sum of a subarray ending
        // exactly here; the answer is its maximum over all indices.
        // Seeding with nums[0] (not 0) makes all-negative inputs come out
        // right: an empty-prefix 0 must not be allowed to win.
        int best = nums[0];
        int current = nums[0];
        for (size_t i = 1; i < nums.size(); i++) {
            int value = nums[i];
            // Extend the best subarray ending at i-1, or start fresh: a
            // negative running sum can only drag down what follows.
            current = current < 0 ? value : current + value;
            if (current > best) {
                best = current;
            }
        }
        return best;
    }
};

File: problems-originals/0001-0100/0053_maximum-subarray/solution_kadane.py
from typing import List, Optional


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Kadane's algorithm: `current` is the best sum of a subarray ending
        # exactly here; every optimal subarray ends somewhere, so the answer
        # is the max of `current` over all indices.
        # Seeding with nums[0] (not 0) is what makes all-negative inputs come
        # out right: an empty-prefix 0 must not be allowed to win.
        best = current = nums[0]
        for value in nums[1:]:
            # Extend the best subarray ending at the previous index, or start
            # fresh: a negative running sum can only drag down what follows.
            current = value if current < 0 else current + value
            if current > best:
                best = current
        return best
